x-pack/plugins/observability_solution/apm/server/utils/es_fields_mappings.ts

```diff
@@ -0,0 +1,951 @@
+  NODEJS_MEMORY_HEAP_ALLOCATED_BYTES,
+  NODEJS_MEMORY_HEAP_USED_BYTES,
+  NODEJS_REQUESTS_ACTIVE,
+  PROCESS_ARGS,
+  PROCESS_PARENT_PID,
+  PROCESS_PID,
+  PROCESS_TITLE,
+  SERVICE_FRAMEWORK_VERSION,
+  SERVICE_LANGUAGE_NAME,
+  SERVICE_RUNTIME_NAME,
+  SERVICE_RUNTIME_VERSION,
+  SERVICE_VERSION,
+  SOURCE_IP,
+  TRANSACTION_DURATION_HISTOGRAM_VALUES,
+  TRANSACTION_DURATION_SUMMARY_SUM,
+  TRANSACTION_DURATION_SUMMARY_VALUE_COUNT,
+  TRANSACTION_SPAN_COUNT_STARTED,
+  URL_DOMAIN,
+  URL_ORIGINAL,
+  URL_PATH,
+  URL_PORT,
+  URL_SCHEME,
+  USER_AGENT_DEVICE_NAME,
+  USER_AGENT_NAME,
+  USER_AGENT_ORIGINAL,
+  USER_AGENT_VERSION,
+} from '@kbn/apm-types';
+import { Transaction } from '../../typings/es_schemas/ui/transaction';
+import { TransactionRaw } from '../../typings/es_schemas/raw/transaction_raw';
+import {
+  WaterfallError,
+  WaterfallSpan,
+  WaterfallTransaction,
+} from '../../common/waterfall/typings';
+import { EventOutcome } from '../../common/event_outcome';
+import { Exception } from '../../typings/es_schemas/raw/error_raw';
+
+type ServiceMetadataIconsRaw = Pick<TransactionRaw, 'kubernetes' | 'cloud' | 'container' | 'agent'>;
+
+const normalizeValue = <T>(field: unknown[] | unknown): T => {
+  return (Array.isArray(field) && field.length > 0 ? field[0] : field) as T;
+};
+
+export const metadataForDependencyMapping = (fields: Partial<Record<string, unknown[]>>) => {
+  return {
+    span: {
+      type: normalizeValue<string>(fields[SPAN_TYPE]),
+      subtype: normalizeValue<string>(fields[SPAN_SUBTYPE]),
+    },
+  };
+};
+
+export const transactionsForDependencySpansMapping = (
+  fields: Partial<Record<string, unknown[]>>
+) => {
+  return {
+    transaction: {
+      id: normalizeValue<string>(fields[TRANSACTION_ID]),
+      name: normalizeValue<string>(fields[TRANSACTION_NAME]),
+      type: normalizeValue<string>(fields[TRANSACTION_TYPE]),
+    },
+  };
+};
+
+export const topDependencySpansMapping = (fields: Partial<Record<string, unknown[]>>) => {
+  return {
+    transaction: {
+      id: normalizeValue<string>(fields[TRANSACTION_ID]),
+    },
+    span: {
+      id: normalizeValue<string>(fields[SPAN_ID]),
+      name: normalizeValue<string>(fields[SPAN_NAME]),
+      duration: {
+        us: normalizeValue<number>(fields[SPAN_DURATION]),
+      },
+    },
+    trace: {
+      id: normalizeValue<string>(fields[TRACE_ID]),
+    },
+    service: {
+      name: normalizeValue<string>(fields[SERVICE_NAME]),
+      environment: normalizeValue<string>(fields[SERVICE_ENVIRONMENT]),
+    },
+    agent: {
+      name: normalizeValue<AgentName>(fields[AGENT_NAME]),
+    },
+    event: {
+      outcome: normalizeValue<EventOutcome>(fields[EVENT_OUTCOME]),
+    },
+    '@timestamp': normalizeValue<string>(fields[AT_TIMESTAMP]),
+  };
+};
+
+// todo(milosz): test with otel APM POC, mapping format wasn't confirmed with `fields`
+export const spanMapping = (fields: Partial<Record<string, unknown[]>>): Span => {
```

Comment:
I noticed that `SpanRaw` has a couple of additional fields not represented here, such as `stacktrace`. Do those need to be represented here?

Reply:
While checking it I found that `get_span` can give completely different response depending on span. When I first created mapping I based it on response from `edge-lite-oblt` cluster, now I checked `otel-apm-e2e-poc` and received different response format. I covered two possible responses in the updated mapping, but I wonder if there are other unknown yet responses for that one. BTW I haven't found `stacktrace` in neither of two.

Reply:
get_span in `otel-apm-e2e-poc`:
<img width="2560" alt="Screenshot 2024-09-25 at 16 52 38" src="https://github.com/user-attachments/assets/98eb5f90-84d1-44e7-99d3-4d715757d826">

get_span in `edge-lite-oblt`:
<img width="2560" alt="Screenshot 2024-09-25 at 16 55 25" src="https://github.com/user-attachments/assets/d7fd934e-2d5f-4344-b4f6-9f1609a3cdd4">